File: sangsoo/0x0D Simulation/13460.cpp
// fail
// https://junbastick.tistory.com/37


#include <bits/stdc++.h>
using namespace std;

struct step{
    int Rx, Ry;
    int Bx, By;
    int Count;
};

char board[11][11];
bool visited[11][11][11][11];
int N, M;
int dx[] = {1, -1, 0, 0};
int dy[] = {0, 0, 1, -1};

void move(int& rx, int& ry, int& distance, int& i){
    while (board[rx+dx[i]][ry+dy[i]] != '#' && board[rx][ry] != 'O'){
        rx += dx[i];
        ry += dy[i];
        distance += 1;
    }
}

void BFS(int Rx, int Ry, int Bx, int By){
    queue<step> q;
    q.push({Rx, Ry, Bx, By, 0});
    visited[Rx][Ry][Bx][By] = true;
    while(!q.empty()){
        int rx = q.front().Rx;
        int ry = q.front().Ry;
        int bx = q.front().Bx;
        int by = q.front().By;
        int count = q.front().Count;
        q.pop();
        if (count >= 10) break;
        for (int i = 0; i < 4; i++){
            int nrx = rx, nry=ry, nbx = bx, nby = by;
            int rc = 0, bc = 0, ncount = count+1;
            move(nrx, nry, rc, i);
            move(nbx, nby, bc, i);
            if(board[nbx][nby] == 'O') continue;
            if(board[nrx][nry] == 'O'){
                cout << ncount;
                return;
            }
            if(nrx == nbx && nry == nby){
                if(rc > bc) nrx -= dx[i], nry -= dy[i];
                else nbx -= dx[i], nby -= dy[i];
            }
            if (visited[nrx][nry][nbx][nby]) continue;
            visited[nrx][nry][nbx][nby] = true;
            q.push({nrx, nry, nbx, nby, ncount});
        }
    }
    cout << -1;
}

void Answer(){
    cin >> N >> M;
    int Rx = 0, Ry = 0, Bx = 0, By = 0;
    for (int i = 0; i < N; i++){
        for (int j = 0; j < M; j++){
            cin >> board[i][j];
            if(board[i][j] == 'R') Rx = i, Ry = j;
            else if (board[i][j] == 'B') Bx = i, By = j;
        }
    }
    BFS(Rx, Ry, Bx, By);
}

int main(){
    ios::sync_with_stdio(0);
    cin.tie(0);
    Answer();
    return 0;
}

File: sangsoo/0x0D Simulation/13460.py
# fail
# https://bgspro.tistory.com/70

from collections import deque
import sys
input =  sys.stdin.readline

n, m = map(int, input().split())
graph = []
for i in range(n):
    graph.append(list(input()))
    for j in range(m):
        if graph[i][j] == 'R':
            rx, ry = i, j
        elif graph[i][j] == 'B':
            bx, by = i, j

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

def bfs(rx, ry, bx, by):
    q = deque()
    q.append((rx, ry, bx, by))
    visited = []
    visited.append((rx, ry, bx, by))
    count = 0
    while q:
        for _ in range(len(q)):
            rx, ry, bx, by = q.popleft()
            if count > 10: # 움직인 횟수가 10회 초과면 -1 출력
                print(-1)
                return
            if graph[rx][ry] == 'O':
                print(count)
                return
            for i in range(4):
                nrx, nry = rx, ry
                while True:
                    nrx += dx[i]
                    nry += dy[i]
                    if graph[nrx][nry] == '#':
                        nrx -= dx[i]
                        nry -= dy[i]
                        break
                    if graph[nrx][nry] == 'O':
                        break
                nbx, nby = bx, by
                while True:
                    nbx += dx[i]
                    nby += dy[i]
                    if graph[nbx][nby] == '#':
                        nbx -= dx[i]
                        nby -= dy[i]
                        break
                    if graph[nbx][nby] == 'O':
                        break
                if graph[nbx][nby] == 'O':
                    continue
                if nrx == nbx and nry == nby:
                    if abs(nrx - rx) + abs(nry - ry) > abs(nbx - bx) + abs(nby - by): # 더 많이 이동한 구슬이 더 늦게 이동한 구슬이므로 늦게 이동한 구슬 한칸 뒤로 이동
                        nrx -= dx[i]
                        nry -= dy[i]
                    else:
                        nbx -= dx[i]
                        nby -= dy[i]
                if (nrx, nry, nbx, nby) not in visited:
                    q.append((nrx, nry, nbx, nby))
                    visited.append((nrx, nry, nbx, nby))
        count += 1
    print(-1)
bfs(rx, ry, bx, by)